gen.go

```diff
@@ -1199,33 +1208,77 @@ func GenTupleEncodersForType(gti *GenTypeInfo, w io.Writer) error {
 }
 
 func emitCborMarshalStructMap(w io.Writer, gti *GenTypeInfo) error {
+	var hasOmitEmpty bool
+	for _, f := range gti.Fields {
+		if f.OmitEmpty {
+			hasOmitEmpty = true
+		}
+	}
+
 	err := doTemplate(w, gti, `func (t *{{ .Name }}) MarshalCBOR(w io.Writer) error {
 	if t == nil {
 		_, err := w.Write(cbg.CborNull)
 		return err
 	}
 
 	cw := cbg.NewCborWriter(w)
+`)
+	if err != nil {
+		return err
+	}
+
+	if hasOmitEmpty {
+		fmt.Fprintln(w, "var emptyFieldCount int")
+		for _, f := range gti.Fields {
+			if f.OmitEmpty {
+				err := doTemplate(w, f, `
+	if t.{{ .Name }} == nil {
```

Comment:
This isn't go's definition of "empty".

> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string. 

You can:

- Do nothing if you don't care.
- Use `reflect.Value.IsZero` (some runtime cost).
- Have a bunch of logic to define the "zeros" for different kinds.
- Define per-field zero values as "variables" in codegen (using `reflect.Zero`).

Reply:
Oh, so, there's an easier way.

```go
var {{.Name}}Zero {{.Type}}
if t.{{.Name}} == {{.Name}}Zero {
    ...
}
```

Reply:
yeah... I was getting kinda lazy since this fix was three layers down the rabbit hole. I'd like to avoid runtime costs so i think generator time type checking is the right thing to do